unpretty: `Unexpected def kind SyntheticCoroutineBody`

<!--
[31mICE[0m: Rustc ./a.rs '-Zdump-mir=all --edition=2018 -ooutputfile -Zdump-mir-dir=dir' 'error: internal compiler error: compiler/rustc_middle/src/mir/pretty.rs:579:14: Unexpected def kind SyntheticCoroutineBody', 'error: internal compiler error: compiler/rustc_middle/src/mir/pretty.rs:579:14: Unexpected def kind SyntheticCoroutineBody'
File: /tmp/im2/a.rs
-->

snippet:
````rust
async fn call_once<T>(f: impl async FnOnce() -> T) -> T {
    f().await
}

async fn async_main() {
    {
        let c = async move || {};

        call_once(c).await;
    }
}


````

Version information
````
rustc 1.82.0-nightly (ac77e88f7 2024-08-28)
binary: rustc
commit-hash: ac77e88f7a84e20311f5518e34c806503d586c1c
commit-date: 2024-08-28
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zdump-mir=all --edition=2018`

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: async closures are unstable
 --> /tmp/icemaker_global_tempdir.7pX8pm1PX1An/rustc_testrunner_tmpdir_reporting.FdI2BZBwLkpb/mvce.rs:1:31
  |
1 | async fn call_once<T>(f: impl async FnOnce() -> T) -> T {
  |                               ^^^^^
  |
  = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
  = help: add `#![feature(async_closure)]` to the crate attributes to enable
  = note: this compiler was built on 2024-08-28; consider upgrading it if it is out of date
  = help: to use an async block, remove the `||`: `async {`

error[E0658]: async closures are unstable
 --> /tmp/icemaker_global_tempdir.7pX8pm1PX1An/rustc_testrunner_tmpdir_reporting.FdI2BZBwLkpb/mvce.rs:7:17
  |
7 |         let c = async move || {};
  |                 ^^^^^
  |
  = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
  = help: add `#![feature(async_closure)]` to the crate attributes to enable
  = note: this compiler was built on 2024-08-28; consider upgrading it if it is out of date
  = help: to use an async block, remove the `||`: `async {`

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.7pX8pm1PX1An/rustc_testrunner_tmpdir_reporting.FdI2BZBwLkpb/mvce.rs:11:2
   |
11 | }
   |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.7pX8pm1PX1An/rustc_testrunner_tmpdir_reporting.FdI2BZBwLkpb/mvce.rs`

error[E0658]: use of unstable library feature 'async_closure'
 --> /tmp/icemaker_global_tempdir.7pX8pm1PX1An/rustc_testrunner_tmpdir_reporting.FdI2BZBwLkpb/mvce.rs:1:37
  |
1 | async fn call_once<T>(f: impl async FnOnce() -> T) -> T {
  |                                     ^^^^^^^^^^^^^
  |
  = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
  = help: add `#![feature(async_closure)]` to the crate attributes to enable
  = note: this compiler was built on 2024-08-28; consider upgrading it if it is out of date

error: internal compiler error: compiler/rustc_middle/src/mir/pretty.rs:579:14: Unexpected def kind SyntheticCoroutineBody

thread 'rustc' panicked at compiler/rustc_middle/src/mir/pretty.rs:579:14:
Box<dyn Any>
stack backtrace:
   0:     0x7e90bb9eb32d - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h7db77851e9ebb774
   1:     0x7e90bc2031d7 - core::fmt::write::h9cfbdc571e0dc2eb
   2:     0x7e90bd1d9c91 - std::io::Write::write_fmt::h3ab32c8206ff30b6
   3:     0x7e90bb9eda0b - std::panicking::default_hook::{{closure}}::h38e1e6aafa3385ea
   4:     0x7e90bb9ed67e - std::panicking::default_hook::h128e065af5edc5f1
   5:     0x7e90bab4ff99 - std[1420b3a934c096b7]::panicking::update_hook::<alloc[e966b72f101d6e3b]::boxed::Box<rustc_driver_impl[bb501fee5761e404]::install_ice_hook::{closure#0}>>::{closure#0}
   6:     0x7e90bb9ee327 - std::panicking::rust_panic_with_hook::h3da31308f8371ddd
   7:     0x7e90bab8a1e1 - std[1420b3a934c096b7]::panicking::begin_panic::<rustc_errors[975dcf607b16fe9a]::ExplicitBug>::{closure#0}
   8:     0x7e90bab7d8a6 - std[1420b3a934c096b7]::sys::backtrace::__rust_end_short_backtrace::<std[1420b3a934c096b7]::panicking::begin_panic<rustc_errors[975dcf607b16fe9a]::ExplicitBug>::{closure#0}, !>
   9:     0x7e90bab78bf6 - std[1420b3a934c096b7]::panicking::begin_panic::<rustc_errors[975dcf607b16fe9a]::ExplicitBug>
  10:     0x7e90bab93401 - <rustc_errors[975dcf607b16fe9a]::diagnostic::BugAbort as rustc_errors[975dcf607b16fe9a]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  11:     0x7e90bb15b064 - rustc_middle[8198af3142fd3a9]::util::bug::opt_span_bug_fmt::<rustc_span[b2ae2de433ea6577]::span_encoding::Span>::{closure#0}
  12:     0x7e90bb1412ea - rustc_middle[8198af3142fd3a9]::ty::context::tls::with_opt::<rustc_middle[8198af3142fd3a9]::util::bug::opt_span_bug_fmt<rustc_span[b2ae2de433ea6577]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  13:     0x7e90bb14119b - rustc_middle[8198af3142fd3a9]::ty::context::tls::with_context_opt::<rustc_middle[8198af3142fd3a9]::ty::context::tls::with_opt<rustc_middle[8198af3142fd3a9]::util::bug::opt_span_bug_fmt<rustc_span[b2ae2de433ea6577]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  14:     0x7e90b8cc8f80 - rustc_middle[8198af3142fd3a9]::util::bug::bug_fmt
  15:     0x7e90bb11ab88 - rustc_middle[8198af3142fd3a9]::mir::pretty::write_mir_intro
  16:     0x7e90bb260977 - rustc_middle[8198af3142fd3a9]::mir::pretty::write_mir_fn::<rustc_mir_transform[bd54f7397a7bba57]::pass_manager::dump_mir_for_pass::{closure#0}>
  17:     0x7e90b998f6d6 - rustc_mir_transform[bd54f7397a7bba57]::mir_promoted
  18:     0x7e90bc6ade54 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::mir_promoted::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  19:     0x7e90bc6ae107 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::VecCache<rustc_span[b2ae2de433ea6577]::def_id::LocalDefId, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  20:     0x7e90bd3000d0 - rustc_query_impl[58736c819e430f36]::query_impl::mir_promoted::get_query_non_incr::__rust_end_short_backtrace
  21:     0x7e90bdae11f9 - rustc_middle[8198af3142fd3a9]::query::plumbing::query_get_at::<rustc_query_system[cbe41465d9b6afb1]::query::caches::VecCache<rustc_span[b2ae2de433ea6577]::def_id::LocalDefId, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>>.cold
  22:     0x7e90bb33e58a - rustc_mir_transform[bd54f7397a7bba57]::coroutine::mir_coroutine_witnesses
  23:     0x7e90bb528972 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::mir_coroutine_witnesses::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>
  24:     0x7e90bc23136e - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefIdCache<rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  25:     0x7e90bb530e7e - rustc_query_impl[58736c819e430f36]::query_impl::mir_coroutine_witnesses::get_query_non_incr::__rust_end_short_backtrace
  26:     0x7e90bcb06d6a - rustc_mir_transform[bd54f7397a7bba57]::mir_drops_elaborated_and_const_checked
  27:     0x7e90bcb05ffd - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>
  28:     0x7e90bc3c9136 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::VecCache<rustc_span[b2ae2de433ea6577]::def_id::LocalDefId, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  29:     0x7e90bc3c8ae5 - rustc_query_impl[58736c819e430f36]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7e90bccc4f2e - rustc_mir_transform[bd54f7397a7bba57]::optimized_mir
  31:     0x7e90bccc3669 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>
  32:     0x7e90bc23136e - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefIdCache<rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  33:     0x7e90bc23091f - rustc_query_impl[58736c819e430f36]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  34:     0x7e90bcbf2a7d - rustc_middle[8198af3142fd3a9]::query::plumbing::query_get_at::<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefIdCache<rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>>
  35:     0x7e90bb15ad2e - <rustc_middle[8198af3142fd3a9]::ty::context::TyCtxt>::coroutine_layout
  36:     0x7e90bca5a762 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached
  37:     0x7e90bca54d06 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  38:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  39:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  40:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7e90bca528cd - <core[d1ee7ab0152fc8ae]::iter::adapters::GenericShunt<core[d1ee7ab0152fc8ae]::iter::adapters::by_ref_sized::ByRefSized<core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<core[d1ee7ab0152fc8ae]::slice::iter::Iter<rustc_middle[8198af3142fd3a9]::ty::VariantDef>, rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached::{closure#13}>>, core[d1ee7ab0152fc8ae]::result::Result<core[d1ee7ab0152fc8ae]::convert::Infallible, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>> as core[d1ee7ab0152fc8ae]::iter::traits::iterator::Iterator>::next
  42:     0x7e90bca56916 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached
  43:     0x7e90bca54d06 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  44:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  45:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  46:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  47:     0x7e90bca52ccb - <core[d1ee7ab0152fc8ae]::iter::adapters::GenericShunt<core[d1ee7ab0152fc8ae]::iter::adapters::by_ref_sized::ByRefSized<core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<core[d1ee7ab0152fc8ae]::slice::iter::Iter<rustc_middle[8198af3142fd3a9]::ty::VariantDef>, rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached::{closure#13}>>, core[d1ee7ab0152fc8ae]::result::Result<core[d1ee7ab0152fc8ae]::convert::Infallible, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>> as core[d1ee7ab0152fc8ae]::iter::traits::iterator::Iterator>::next
  48:     0x7e90bca56916 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached
  49:     0x7e90bca54d06 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  50:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  51:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  52:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  53:     0x7e90bca51e9d - rustc_middle[8198af3142fd3a9]::query::plumbing::query_get_at::<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>>
  54:     0x7e90bca550af - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  55:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  56:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  57:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  58:     0x7e90bca51e9d - rustc_middle[8198af3142fd3a9]::query::plumbing::query_get_at::<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>>
  59:     0x7e90bb96923a - <rustc_middle[8198af3142fd3a9]::ty::layout::LayoutCx<rustc_middle[8198af3142fd3a9]::ty::context::TyCtxt> as rustc_middle[8198af3142fd3a9]::ty::layout::LayoutOf>::spanned_layout_of
  60:     0x7e90bb9553da - <core[d1ee7ab0152fc8ae]::iter::adapters::GenericShunt<core[d1ee7ab0152fc8ae]::iter::adapters::by_ref_sized::ByRefSized<core[d1ee7ab0152fc8ae]::iter::adapters::chain::Chain<core[d1ee7ab0152fc8ae]::iter::adapters::chain::Chain<core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<core[d1ee7ab0152fc8ae]::iter::adapters::copied::Copied<core[d1ee7ab0152fc8ae]::slice::iter::Iter<rustc_middle[8198af3142fd3a9]::ty::Ty>>, rustc_ty_utils[106d4f25596debe8]::layout::coroutine_layout::{closure#2}>, core[d1ee7ab0152fc8ae]::iter::sources::once::Once<core[d1ee7ab0152fc8ae]::result::Result<rustc_target[a31bdeae75b94f7f]::abi::Layout, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>>>, core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<rustc_index[bb3120484daaecec]::bit_set::BitIter<rustc_middle[8198af3142fd3a9]::mir::query::CoroutineSavedLocal>, rustc_ty_utils[106d4f25596debe8]::layout::coroutine_layout::{closure#1}>>>, core[d1ee7ab0152fc8ae]::result::Result<core[d1ee7ab0152fc8ae]::convert::Infallible, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>> as core[d1ee7ab0152fc8ae]::iter::traits::iterator::Iterator>::next
  61:     0x7e90bca5c2de - rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached
  62:     0x7e90bca54d06 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  63:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  64:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  65:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  66:     0x7e90bca528cd - <core[d1ee7ab0152fc8ae]::iter::adapters::GenericShunt<core[d1ee7ab0152fc8ae]::iter::adapters::by_ref_sized::ByRefSized<core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<core[d1ee7ab0152fc8ae]::slice::iter::Iter<rustc_middle[8198af3142fd3a9]::ty::VariantDef>, rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached::{closure#13}>>, core[d1ee7ab0152fc8ae]::result::Result<core[d1ee7ab0152fc8ae]::convert::Infallible, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>> as core[d1ee7ab0152fc8ae]::iter::traits::iterator::Iterator>::next
  67:     0x7e90bca56916 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached
  68:     0x7e90bca54d06 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  69:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  70:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  71:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  72:     0x7e90bca52ccb - <core[d1ee7ab0152fc8ae]::iter::adapters::GenericShunt<core[d1ee7ab0152fc8ae]::iter::adapters::by_ref_sized::ByRefSized<core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<core[d1ee7ab0152fc8ae]::slice::iter::Iter<rustc_middle[8198af3142fd3a9]::ty::VariantDef>, rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached::{closure#13}>>, core[d1ee7ab0152fc8ae]::result::Result<core[d1ee7ab0152fc8ae]::convert::Infallible, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>> as core[d1ee7ab0152fc8ae]::iter::traits::iterator::Iterator>::next
  73:     0x7e90bca56916 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached
  74:     0x7e90bca54d06 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  75:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  76:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  77:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  78:     0x7e90bca51e9d - rustc_middle[8198af3142fd3a9]::query::plumbing::query_get_at::<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>>
  79:     0x7e90bb96923a - <rustc_middle[8198af3142fd3a9]::ty::layout::LayoutCx<rustc_middle[8198af3142fd3a9]::ty::context::TyCtxt> as rustc_middle[8198af3142fd3a9]::ty::layout::LayoutOf>::spanned_layout_of
  80:     0x7e90bb9553da - <core[d1ee7ab0152fc8ae]::iter::adapters::GenericShunt<core[d1ee7ab0152fc8ae]::iter::adapters::by_ref_sized::ByRefSized<core[d1ee7ab0152fc8ae]::iter::adapters::chain::Chain<core[d1ee7ab0152fc8ae]::iter::adapters::chain::Chain<core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<core[d1ee7ab0152fc8ae]::iter::adapters::copied::Copied<core[d1ee7ab0152fc8ae]::slice::iter::Iter<rustc_middle[8198af3142fd3a9]::ty::Ty>>, rustc_ty_utils[106d4f25596debe8]::layout::coroutine_layout::{closure#2}>, core[d1ee7ab0152fc8ae]::iter::sources::once::Once<core[d1ee7ab0152fc8ae]::result::Result<rustc_target[a31bdeae75b94f7f]::abi::Layout, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>>>, core[d1ee7ab0152fc8ae]::iter::adapters::map::Map<rustc_index[bb3120484daaecec]::bit_set::BitIter<rustc_middle[8198af3142fd3a9]::mir::query::CoroutineSavedLocal>, rustc_ty_utils[106d4f25596debe8]::layout::coroutine_layout::{closure#1}>>>, core[d1ee7ab0152fc8ae]::result::Result<core[d1ee7ab0152fc8ae]::convert::Infallible, &rustc_middle[8198af3142fd3a9]::ty::layout::LayoutError>> as core[d1ee7ab0152fc8ae]::iter::traits::iterator::Iterator>::next
  81:     0x7e90bca5c2de - rustc_ty_utils[106d4f25596debe8]::layout::layout_of_uncached
  82:     0x7e90bca54d06 - rustc_ty_utils[106d4f25596debe8]::layout::layout_of
  83:     0x7e90bca54c91 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>
  84:     0x7e90bca53ed3 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::DefaultCache<rustc_middle[8198af3142fd3a9]::ty::ParamEnvAnd<rustc_middle[8198af3142fd3a9]::ty::Ty>, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  85:     0x7e90bca53b6d - rustc_query_impl[58736c819e430f36]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  86:     0x7e90b965a82d - <rustc_mir_transform[bd54f7397a7bba57]::known_panics_lint::KnownPanicsLint as rustc_mir_transform[bd54f7397a7bba57]::pass_manager::MirLint>::run_lint
  87:     0x7e90bc208684 - rustc_mir_transform[bd54f7397a7bba57]::run_analysis_to_runtime_passes
  88:     0x7e90bcb066ba - rustc_mir_transform[bd54f7397a7bba57]::mir_drops_elaborated_and_const_checked
  89:     0x7e90bcb05ffd - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>
  90:     0x7e90bc3c9136 - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::VecCache<rustc_span[b2ae2de433ea6577]::def_id::LocalDefId, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  91:     0x7e90bc3c8ae5 - rustc_query_impl[58736c819e430f36]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  92:     0x7e90bcc52bfe - rustc_interface[7fb29fbff936f0be]::passes::run_required_analyses
  93:     0x7e90bcc4871e - rustc_interface[7fb29fbff936f0be]::passes::analysis
  94:     0x7e90bcc486f1 - rustc_query_impl[58736c819e430f36]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[58736c819e430f36]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 1usize]>>
  95:     0x7e90bd167d6e - rustc_query_system[cbe41465d9b6afb1]::query::plumbing::try_execute_query::<rustc_query_impl[58736c819e430f36]::DynamicConfig<rustc_query_system[cbe41465d9b6afb1]::query::caches::SingleCache<rustc_middle[8198af3142fd3a9]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[58736c819e430f36]::plumbing::QueryCtxt, false>
  96:     0x7e90bd167acf - rustc_query_impl[58736c819e430f36]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  97:     0x7e90bcfc092a - rustc_interface[7fb29fbff936f0be]::interface::run_compiler::<core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>, rustc_driver_impl[bb501fee5761e404]::run_compiler::{closure#0}>::{closure#1}
  98:     0x7e90bd0509d0 - std[1420b3a934c096b7]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[7fb29fbff936f0be]::util::run_in_thread_with_globals<rustc_interface[7fb29fbff936f0be]::util::run_in_thread_pool_with_globals<rustc_interface[7fb29fbff936f0be]::interface::run_compiler<core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>, rustc_driver_impl[bb501fee5761e404]::run_compiler::{closure#0}>::{closure#1}, core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>>::{closure#0}, core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>>
  99:     0x7e90bd05103a - <<std[1420b3a934c096b7]::thread::Builder>::spawn_unchecked_<rustc_interface[7fb29fbff936f0be]::util::run_in_thread_with_globals<rustc_interface[7fb29fbff936f0be]::util::run_in_thread_pool_with_globals<rustc_interface[7fb29fbff936f0be]::interface::run_compiler<core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>, rustc_driver_impl[bb501fee5761e404]::run_compiler::{closure#0}>::{closure#1}, core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>>::{closure#0}, core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[d1ee7ab0152fc8ae]::result::Result<(), rustc_span[b2ae2de433ea6577]::ErrorGuaranteed>>::{closure#1} as core[d1ee7ab0152fc8ae]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
 100:     0x7e90bd0513ab - std::sys::pal::unix::thread::Thread::new::thread_start::h2097b27b23ed513c
 101:     0x7e90be80139d - <unknown>
 102:     0x7e90be88649c - <unknown>
 103:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.82.0-nightly (ac77e88f7 2024-08-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z dump-mir=all -Z dump-mir-dir=dir

query stack during panic:
#0 [mir_promoted] promoting constants in MIR for `async_main::{closure#0}::{closure#0}::{closure#0}::{closure#0}`
#1 [mir_coroutine_witnesses] coroutine witness types for `async_main::{closure#0}::{closure#0}::{closure#0}::{closure#0}`
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [mir_promoted] promoting constants in MIR for `async_main::{closure#0}::{closure#0}::{closure#0}::{closure#0}`
#1 [mir_coroutine_witnesses] coroutine witness types for `async_main::{closure#0}::{closure#0}::{closure#0}::{closure#0}`

-->

--- a/compiler/rustc_middle/src/ty/util.rs
+++ b/compiler/rustc_middle/src/ty/util.rs
@@ -571,7 +571,10 @@ impl<'tcx> TyCtxt<'tcx> {
     /// Returns `true` if `def_id` refers to a definition that does not have its own
     /// type-checking context, i.e. closure, coroutine or inline const.
     pub fn is_typeck_child(self, def_id: DefId) -> bool {
-        matches!(self.def_kind(def_id), DefKind::Closure | DefKind::InlineConst)
+        matches!(
+            self.def_kind(def_id),
+            DefKind::Closure | DefKind::InlineConst | DefKind::SyntheticCoroutineBody
+        )
     }
 
     /// Returns `true` if `def_id` refers to a trait (i.e., `trait Foo { ... }`).

--- a/tests/ui/async-await/async-closures/debuginfo-by-move-body.rs
+++ b/tests/ui/async-await/async-closures/debuginfo-by-move-body.rs
@@ -1,0 +1,19 @@
+//@ aux-build:block-on.rs
+//@ edition: 2021
+//@ build-pass
+//@ compile-flags: -Cdebuginfo=2
+
+#![feature(async_closure)]
+
+extern crate block_on;
+
+async fn call_once(f: impl async FnOnce()) {
+    f().await;
+}
+
+pub fn main() {
+    block_on::block_on(async {
+        let async_closure = async move || {};
+        call_once(async_closure).await;
+    });
+}

